includes/class-wcsatt-order.php

```diff
@@ -0,0 +1,153 @@
+		$scheme_key = null;
+		// Because who trusts WC 3.0 for back-compat?
+		if ( WCS_ATT_Core_Compatibility::is_wc_version_gte_2_7() ) {
+			if ( $order_item->meta_exists( '_wcsatt_scheme' ) ) {
+				$scheme_key = $order_item->get_meta( '_wcsatt_scheme', true );
+				$scheme_key = 0 === absint( $scheme_key ) ? false : strval( $scheme_key );
+			// Backwards compatibility with v1.
+			} elseif ( $order_item->meta_exists( '_wcsatt_scheme_id' ) ) {
+				$scheme_key = $order_item->get_meta( '_wcsatt_scheme_id', true );
+				$scheme_key = 0 === absint( $scheme_key ) ? false : strval( $scheme_key );
+			}
+		} else {
+			if ( isset( $order_item[ 'wcsatt_scheme' ] ) ) {
+				$scheme_key = 0 === absint( $order_item[ 'wcsatt_scheme' ] ) ? false : strval( $order_item[ 'wcsatt_scheme' ] );
+			// Backwards compatibility with v1.
+			} elseif ( isset( $order_item[ 'wcsatt_scheme_id' ] ) ) {
+				$scheme_key = 0 === absint( $order_item[ 'wcsatt_scheme_id' ] ) ? false : strval( $order_item[ 'wcsatt_scheme_id' ] );
+			}
+		}
+		return $scheme_key;
+	}
+
+	/*
+	|--------------------------------------------------------------------------
+	| Hooks
+	|--------------------------------------------------------------------------
+	*/
+
+	/**
+	 * Attempts to restore subscription data when creating a cart item using an order item as reference.
+	 *
+	 * @param  array     $cart_item
+	 * @param  array     $order_item
+	 * @param  WC_Order  $order
+	 * @return array
+	 */
+	public static function restore_cart_item_from_order_item( $cart_item, $order_item, $order ) {
+		if ( null !== ( $scheme_key = self::get_subscription_scheme( $order_item ) ) ) {
+			$cart_item[ 'wcsatt_data' ] = array(
+				'active_subscription_scheme' => $scheme_key
+			);
+		}
+		return $cart_item;
+	}
+
+	/**
+	 * Attempts to restore the subscription state of a product instantiated using an order item as reference.
+	 *
+	 * @param  WC_Product  $product
+	 * @param  array       $order_item
+	 * @return WC_Product
+	 */
+	public static function restore_product_from_order_item( $product, $order_item ) {
+		if ( null !== ( $scheme_key = self::get_subscription_scheme( $order_item ) ) ) {
+			WCS_ATT_Product::set_subscription_scheme( $product, $scheme_key );
+		}
+		return $product;
+	}
+
+	/**
+	 * Stores the scheme key against the order item.
+	 * Used for reconstructing the scheme when reordering, resubscribing, etc - @see 'WCS_ATT_Cart::add_cart_item_data'.
+	 *
+	 * @param  WC_Order_Item  $order_item
+	 * @param  string         $cart_item_key
+	 * @param  array          $cart_item
+	 * @return void
+	 */
+	public static function save_subscription_scheme_meta( $order_item, $cart_item_key, $cart_item ) {
+
+		$scheme_key = WCS_ATT_Cart::get_subscription_scheme( $cart_item );
+
+		if ( null !== $scheme_key ) {
+			$scheme_key = false === $scheme_key ? '0' : $scheme_key;
+			$order_item->add_meta_data( '_wcsatt_scheme', $scheme_key, true );
```

Comment:
The approach I'd follow would be to start fresh with data/hook names in WCS 3.0, and expect existing SATT users to install a WCS 3.0-compatible version that will:

- migrate their data; and
- include a hook deprecator.

For details, see https://github.com/Prospress/woocommerce-subscribe-all-the-things/pull/136#discussion_r125601822